include/drivers/video.h

```diff
@@ -0,0 +1,301 @@
+/**
+ * @file
+ *
+ * @brief Public APIs for Video.
+ */
+
+/*
+ * Copyright (c) 2019 Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_VIDEO_H_
+#define ZEPHYR_INCLUDE_VIDEO_H_
+
+/**
+ * @brief Video Interface
+ * @defgroup video_interface Video Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <device.h>
+#include <stddef.h>
+#include <zephyr.h>
+
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+#define VIDEO_CAP_CAPTURE	0x00000001 /* Video capture device/buffer */
+#define VIDEO_CAP_OUTPUT	0x00000002 /* Video output device/buffer */
+#define VIDEO_CAP_M2M		0x00000004 /* Video memory to memory device */
+
+/**
+ * @brief video format structure
+ *
+ * Used to configure frame format.
+ *
+ * @param pixelformat is the fourcc pixel format value.
+ * @param width is the frame width in pixels.
+ * @param height is the frame height in pixels.
+ * @param pitch is the line stride, the number of bytes that needs to be added
+ *    to the address in the first pixel of a row in order to go to the address
+ *    of the first pixel of the next row (>=width).
+ */
+struct video_format {
+	u32_t pixelformat;
+	u32_t width;
+	u32_t height;
+	u32_t pitch;
+};
+
+/**
+ * @brief video buffer structure
+ *
+ * Represent a video frame.
+ *
+ * @param driver_data is a pointer to driver specific data.
+ * @param frame is a pointer to the frame buffer containing pixels.
+ * @param bytesused is the number of bytes occupied by the data in the buffer.
+ * @param type is the type of buffer (capture, output).
+ * @param fmt is the video frame format.
+ */
+struct video_buffer {
+	void *driver_data;
+	void *frame;
+	u32_t bytesused;
+	u32_t type;
+	struct video_format fmt;
```

Comment:
Video format will be constant in whole stream. IMHO this should be property of the stream.

Reply:
I'm not sure about that, video buffer is a unitary element and should be self-described, and be able to exist out of a stream. E.g. If I do so post-processing on a buffer that is not part of a stream or whose stream has been released... But this also depends on how the concept of 'stream' will be implemented, If possible, I would like to go with a simple API for now, and discuss this as API improvement.

Reply:
That would be true if you are taking image from the camera. But video is a series of images with the same properties.

As abstraction allows for series of video frames having different format and resolution, you will have to handle that in all devices - for example by taking the values from first frame and checking in any other enqueued has the same properties.

Summarising, we have two options:
- Abstracting images (which might be later converted to video stream)
- Abstracting video (which is a stream of frames/fields).

Note, that image is unitary element, that can be processed without reference to other images. In case of video this is no longer true. While uncompressed _frames_ has this property, neither _field_ or _compressed_frame/field_  could be shown/processed  independently from the others frames/fields.

Reply:
@pizi-nordic, So I finally had some time to review different video frameworks and changed my mind on this, we can indeed see the video buffer as the transport mechanism only, without any requirement on content. The format itself is defined by the endpoint/port.